pkg/services/ngalert/eval/eval.go

```diff
@@ -430,122 +446,151 @@ func datasourceUIDsToRefIDs(refIDsToDatasourceUIDs map[string]string) map[string
 //   - Nonzero (e.g 1.2, NaN) results in Alerting.
 //   - nil results in noData.
 //   - unsupported Frame schemas results in Error.
-func evaluateExecutionResult(execResults ExecutionResults, ts time.Time) Results {
+func evaluateExecutionResult(execResults ExecutionResults, ts time.Time) EvaluationResult {
 	evalResults := make([]Result, 0)
 
-	appendErrRes := func(e error) {
-		evalResults = append(evalResults, NewResultFromError(e, ts, time.Since(ts)))
-	}
-
-	appendNoData := func(labels data.Labels) {
-		evalResults = append(evalResults, Result{
-			State:              NoData,
-			Instance:           labels,
-			EvaluatedAt:        ts,
-			EvaluationDuration: time.Since(ts),
-		})
+	result := EvaluationResult{
+		EvaluatedAt:        ts,
+		EvaluationDuration: time.Since(ts),
 	}
 
 	if execResults.Error != nil {
-		appendErrRes(execResults.Error)
-		return evalResults
+		result.Error = execResults.Error
+		return result
 	}
 
 	if len(execResults.NoData) > 0 {
-		noData := datasourceUIDsToRefIDs(execResults.NoData)
-		for datasourceUID, refIDs := range noData {
-			appendNoData(data.Labels{
-				"datasource_uid": datasourceUID,
-				"ref_id":         strings.Join(refIDs, ","),
-			})
-		}
-		return evalResults
+		result.NoData = &NoDataResult{DatasourceToRefID: datasourceUIDsToRefIDs(execResults.NoData)}
```

Comment:
I think this has the same issue where a Classic Condition that checks for `HasNoValue` will be `No Data` instead of `Firing`?

![Screenshot 2023-01-12 at 14 31 50](https://user-images.githubusercontent.com/85952834/212093859-8b3ae88d-d593-4a40-a322-8e924032cba7.png)
![Screenshot 2023-01-12 at 14 32 00](https://user-images.githubusercontent.com/85952834/212093895-3859b921-358e-4a83-b635-43549b2ab9d2.png)